blog/_src/posts/2017-11-XX-adding-refinement-types.md

```diff
@@ -0,0 +1,292 @@
+
+
+ `vector-copy!` is similar, but instead of returning a fresh vector, the elements in the specified range are copied from the vector `src` into the vector `dest`:
+
+```racket
+(: vector-copy!
+   (All (A) (-> ([dest : (Vectorof A)]
+                 [dest-start : Natural]
+                 [src : (Vectorof A)]
+                 [src-start : Natural]
+                 [src-end : Natural])
+                #:pre (dest dest-start src src-start src-end)
+                (and (<= src-start src-end (vector-length src))
+                     (<= dest-start (vector-length dest))
+                     (<= (- src-end src-start)
+                         (- (vector-length dest) dest-start)))
+                Void)))
+(define (vector-copy! dest dest-start src src-start src-end)
+  (define count (- src-end src-start))
+  (let loop! ([i : Natural 0])
+    (when (< i count)
+      (define a (safe-vector-ref src (+ src-start i)))
+      (safe-vector-set! dest (+ dest-start i) a)
+      (loop! (+ 1 i)))))
+```
+
+`vector-copy` has a more complicated precondition because it includes constraints that must hold with respect to both vectors `src` and `dest`, i.e. the range in the `src` vector [`src-start`,`src-end`) must be a valid range of elements, and the destination vector `dest` must have a corresponding valid range [`dest-start`,`(+ dest-start (- src-end src-start))`).
+
+
+## Final Notes
+
+### The above examples and Racket 6.11
+
+To make the above examples work on Racket v6.11, the following snippet---which axiomatizes a more specific type for `make-vector`---is required:
+
+```racket
+(unsafe-require/typed 
+ typed/racket/base
+ [make-vector
+  (All (A) (-> ([size : Natural]
+                [val : A])
+               (Refine [v : (Vectorof A)]
+                       (= size (vector-length v)))))])
+```
+
+This oversight was fixed shortly after the release and the additional type information for usages of `make-vector` should be automatically available when the `#:with-refinements` option is used in a module.
+
+### Linear integer arithmetic
+
+Typed Racket today only reasons about linear integer constraints, i.e. arithmetic constraints expressible with `<=`, `and`, and `or` over [linear combinations](https://en.wikipedia.org/wiki/Linear_combination) of select program terms. If the specification you wish to encode in types requires non-linear arithmetic, Typed Racket will not be able to verify those constraints.
+
+### Dependent function type limitations
+
+Dependent function types currently only support mandatory positional arguments. In the future we plan to add optional, rest, and keyword argument support.
+
+Also, currently functions with dependencies between arguments and/or preconditions cannot be used within a `case->` form.
+
+### Bugs
+
+If you have found a bug or have an idea for how these features might be more useful, feel free to open an issue on Typed Racket's [github repo](https://github.com/racket/typed-racket/issues).
```

Comment:
What's the story for untyped modules that interact with typed modules that use refinement types?